rs/replicated_state/src/canister_state/queues/input_schedule.rs

```diff
@@ -0,0 +1,325 @@
+use ic_validate_eq_derive::ValidateEq;
+use std::collections::{BTreeSet, VecDeque};
+use std::convert::TryFrom;
+
+#[cfg(test)]
+mod tests;
+
+/// Round-robin schedule for canister inputs, across ingress, local senders and
+/// remote senders; as well as within local senders and remote senders.
+///
+/// We define three buckets of queues ("sources"): messages from canisters on
+/// the same subnet ("local senders"); ingress; and messages from canisters on
+/// other subnets ("remote senders"). Each time an input is popped or skipped,
+/// we rotate round-robin between the three sources. We also rotate round-robin
+/// over the canisters in the local sender and remote serder buckets when we pop
+/// or skip messages from those sources.
+#[derive(Clone, Debug, Default, PartialEq, Eq, ValidateEq)]
+pub(super) struct InputSchedule {
+    /// The input source (local senders, ingress or remote senders) at the front
+    /// of the schedule.
+    ///
+    /// Used to implement round-robin rotation over input sources.
+    #[validate_eq(Ignore)]
+    next_input_source: InputSource,
+
+    /// FIFO queue of local subnet sender canister IDs ensuring round-robin
+    /// consumption of input messages. All local senders with non-empty queues
+    /// are scheduled.
+    ///
+    /// We rely on `ReplicatedState::canister_states` to decide whether a canister
+    /// is local or not. This test is subject to race conditions (e.g. if the sender
+    /// has just been deleted), meaning that the separation into local and remote
+    /// senders is best effort.
+    local_sender_schedule: VecDeque<CanisterId>,
+
+    /// FIFO queue of remote subnet sender canister IDs ensuring round-robin
+    /// consumption of input messages. All remote senders with non-empty queues
+    /// are scheduled.
+    ///
+    /// We rely on `ReplicatedState::canister_states` to decide whether a canister
+    /// is local or not. This test is subject to race conditions (e.g. if the sender
+    /// has just been deleted), meaning that the separation into local and remote
+    /// senders is best effort.
+    remote_sender_schedule: VecDeque<CanisterId>,
+
+    /// Set of all senders enqueued in either `local_sender_schedule` or
+    /// `remote_sender_schedule`, to ensure that a sender is enqueued at most once.
+    ///
+    /// We cannot rely on the input queue going from empty to non-empty as the only
+    /// condition for whether to enqueue a sender, because the contents of the queue
+    /// may have been dropped (expired or shed) while the sender was already
+    /// enqueued.
+    scheduled_senders: BTreeSet<CanisterId>,
+}
+
+impl InputSchedule {
+    /// Returns the current input source, without advancing it (e.g. for `peek()`).
+    pub(super) fn input_source(&self) -> InputSource {
+        self.next_input_source
+    }
+
+    /// Returns the current input source and advances to the next one.
+    pub(super) fn next_input_source(&mut self) -> InputSource {
```

Comment:
As far as I can tell these functions are used in tests only at the moment. Is this intended?

Reply:
Yes, I didn't want to rewrite the old `CanisterQueues` to make use of this, only to then immediately throw away the new code in #879.

I did rebase #879 on top of this, though. So the new `CanisterQueues` will be relying on this instead of rolling its own input schedule implementation.